### ISSUE TITLE
Override template Dockerfile or any file in the template.

### ISSUE THREAD

**Issue by japtain-cack** [openfaas/faas-cli#947]
## My actions before raising this issue
- [x] Followed the [troubleshooting guide](https://docs.openfaas.com/deployment/troubleshooting/)
- [x] Read/searched [the docs](https://docs.openfaas.com/)
- [x] Searched [past issues](/issues)

## Why do you need this?
Template files should not be pushed to git, and they should not need to be altered directly. Templates should be, well, templates. Anyway, what I'm looking to do is supply my own Docker file and index.py, as I had to alter them slightly.

I always have to modify the dockerfile, since the default one installs the python dependencies improperly somehow. I always get `ModuleNotFoundError`s for ,anything installed in the `src/requirements.txt`. I've never found a good solution except install the `function/requirements.txt` as root. 

I would like to copy the template file into my project, either in the project root, urnder `/src/`, or something, and have openfaas automatically overwrite/use this file instead of the one in the template. I have already tried various things to see if this is possible, but no dice.

## Who is this for?

Me, you, everyone.

What company is this for? Are you listed in the [ADOPTERS.md](https://github.com/openfaas/faas/blob/master/ADOPTERS.md) file?

My personal kubernetes cluster.

## Expected Behaviour

Supply a custom Dockerfile, or index.py in my project root or subdirectory '(/src/|/overrides/)`, and have openfaas use that, rather then the one in the template directory.

## Current Behaviour

Must modify the template directory files, which shouldn't be submitted to git, so you must either add an exception for that one template in your `.gitignore`, and commit it, or use a script to overwrite it manually.

Overwriting your template files prevents them from being updated when a new template is pulled/updated.

## Are you a GitHub Sponsor (Yes/No?)

Check at: https://github.com/sponsors/openfaas
- [ ] Yes
- [x] No

## List All Possible Solutions and Workarounds

No workaround to my knowledge. See above explanations for details.

## Which Solution Do You Recommend?

n/a

## Steps to Reproduce (for bugs)

1. Put a Dockerfile in your root or `/src` directory.
2. Run `faas publish`.
3. Notice custom Dockerfile is ignored
4. Call my therapist.

## Context
 

Not really affecting anything but my sanity. After working over 10 years for tech companies, I don't have much left,.

## Your Environment

* FaaS-CLI version ( Full output from: `faas-cli version` ):

```shell
faas-cli version
  ___                   _____           ____
 / _ \ _ __   ___ _ __ |  ___|_ _  __ _/ ___|
| | | | '_ \ / _ \ '_ \| |_ / _` |/ _` \___ \
| |_| | |_) |  __/ | | |  _| (_| | (_| |___) |
 \___/| .__/ \___|_| |_|_|  \__,_|\__,_|____/
      |_|

CLI:
 commit:  0074051aeb837f5f160ee8736341460468b5c190
 version: 0.15.4

Gateway
 uri:     https://fn.mimir-tech.org
 version: 0.23.0
 sha:     88eea5f62e1d6a1812265793eacee6213b74af80


Provider
 name:          openfaas-operator
 orchestration: kubernetes
 version:       0.15.1
 sha:           f3309e4bcef20b6a381792d638294265d9e9aa0f
```

* Docker version `docker version` (e.g. Docker 17.0.05 ): Kubernetes: v1.23.6+k3s1

* Are you using OpenFaaS on Kubernetes or [faasd](https://github.com/openfaas/faasd)? Kubernetes

* Operating System and version (e.g. Linux, Windows, MacOS): Linux Ubuntu 20.04 (tooling/cli vm)

* Code example or link to GitHub repo or gist to reproduce problem: https://gitlab.com/japtain_cack/function_grafantfy

* Other diagnostic information / logs from [troubleshooting guide](https://docs.openfaas.com/deployment/troubleshooting): n/a

**Comment by alexellis** [comment]
Hi @japtain-cack thanks for your interest in openfaas. I'm moving this issue to faas-cli.

Since this seems like technical support, we will need you to provide a complete code example and instructions on how to reproduce any errors.

Unfortunately the instructions left in "steps to reproduce" are not something we can use to do that.

Alex

**Comment by japtain-cack** [comment]
Here is my working project. I was more/less looking for a quick solution to replace any template file with a custom one. However, I believe the expected behavior is to clone the template, then map that as the template in the `stack.yaml` file, which is what I've done.

# Desired build process
* `faas new hello-world --lang python3-flask-debian`
* dockerfile, index.py, and the other files are copied over, just like the handler.py, requirements.txt, etc.
* I can edit dockerfile/index.py and handler.py as needed.
* `faas publish --platforms linux/arm64,linux/amd64 --reset-qemu`
* Project is built and pushed with my edited Dockerfile, index.py, handler.py, etc.
* The template is never committed to this repo because it's a URL, so it can be maintained outside this repo.

# Current behavior
* `faas new hello-world --lang python3-flask-debian`
* delete all the template folders I don't need in the `hello-world/templates/` directory.
* Change the stack.yaml `configuration.templates[0].source` to the project's `hello-world/templates/` dir.
* I can edit dockerfile/index.py in the tempalte directory.
* Then I edit the handler.py in my `hello-world/src/` dir as needed.
* `faas publish --platforms linux/arm64,linux/amd64 --reset-qemu`
* Project is built and pushed with my edited Dockerfile, index.py, handler.py, etc.
* Have to commit and maintain template in this repo (undesired).

Mainly, I'm just trying to clarify what the expected/recommended process is for setting up a project, when you have to modify the dockerfile for instance. You can supply a URL as a template source, that way you can get updates to the template files. Or that's the way I understood it primarily. However, if the template isn't perfect for your use case, there is no good way to override certain files in the template. I have to copy the template to my project folder, make the changes to my dockerfile. Then change the `configuration.templates[0].source` in `stack.yml` to a folder in your project (or another repo I suppose).

Basically, it would be awesome to have a builtin way to override template files, and still have that template source url in there. If that functionality is already provided in some way, I couldn't figure it out. 

# My project and current solution (see **Current behavior** above)
https://gitlab.com/japtain_cack/function_grafantfy

**Comment by alexellis** [comment]
Hi @japtain-cack 

Thanks for your interest in OpenFaaS.

Can you create a new issue and explain:

1) What you're trying to do
2) Why
3) Who this is for

Thanks, feel free to include some of what you've written above too.

Alex
